File: Polynomial Regression/PolynomialRegression.R
# Poly Regression

## Beginning of Data Preprocessing ##
dataset = read.csv('Position_Salaries.csv')
# Reset the dataset because we don't need the first column
dataset = dataset[2:3] # Ignore first column
## END of Data Preprocessing ##

## Fitting Linear regression to the dataset
lin_reg = lm(Salary ~ ., data = dataset)

## FItting Poly Regression to the dataset
# Add column into the dataset
dataset$Level2 = (dataset$Level)^2
dataset$Level3 = (dataset$Level)^3
dataset$Level4 = (dataset$Level)^4
poly_reg = lm(Salary ~., data = dataset)

# Visualizing the Linear Regression results
# install.packages('ggplot2) if you dont have ggplot2 installed
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(lin_reg, newdata = dataset)),
            colour = 'blue') + 
  ggtitle('Salary vs Level (Linear Regression)')+
  xlab('Level')+
  ylab('Salary')

# Visualizing the Poly Regression results
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             colour = 'red') + 
  geom_line(aes(x = dataset$Level, y = predict(poly_reg, newdata = dataset)),
            colour = 'blue') + 
  ggtitle('Salary vs Level (Poly Regression)')+
  xlab('Level')+
  ylab('Salary')

# Predict new result with Linear Regression, lin_reg
y_pred = predict(lin_reg, data.frame(Level = 6.5))

# Predict new result with Poly Reg, poly_reg
y_poly_red = predict(poly_reg, data.frame(Level = 6.5,
                                          Level2= 6.5^2,
                                          Level3= 6.5^3,
                                          Level4=6.5^4))

File: Polynomial Regression/polynomialRegression.py
# Polynomial Regression
# The machine is the polynomial regression model
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd


## Beginning of Data Preprocessing ##
# Import the dataset
# prepare the dataset with Matrix X, Y indendent and dependent variables
dataset = pd.read_csv("Position_Salaries.csv")

X = dataset.iloc[:, [1]].values 
# We dont need X[:,0]; I can also use 1:2 instead of [1] to make sure X is a matrix
Y = dataset.iloc[:,-1].values 

# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
lin_reg = LinearRegression()
lin_reg.fit(X, Y)


##  Fitting Polynomial Regression to the data set ##
# Import polynomial class, PolynomialFeatures.(simply add polynomial terms)
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 4)
X_poly = poly_reg.fit_transform(X) 
#Fit our object to X then transform X to X_poly
# It automatically add the column of 1's to include the X0
# Fit X_poly into the linear regressor, lin_reg_2.
# X_poly (consist of 2 independent varaibles X, X^2) because of degree = 2
# WE CAN IMPROVE THIS MODEL BY ADDING THE DEGREE
lin_reg_2 = LinearRegression()
lin_reg_2.fit(X_poly, Y)

# Visualizing the Linear Regression Results
plt.scatter(X, Y, color = 'red')
plt.plot(X, lin_reg.predict(X), 'blue') #USe lin_reg.predict
plt.title('Position level VS Salary (Linear Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')
#Visualizing the Poly Reg Ressults
X_grid = np.arange(min(X), max(X),0.1)
X_grid = X_grid.reshape((len(X_grid)),1) #More Continuous Curve
plt.scatter(X, Y, color = 'red')
plt.plot(X, lin_reg_2.predict(poly_reg.fit_transform(X)), 'blue') #USe lin_reg.predict
plt.title('Position level VS Salary (Poly Regression)')
plt.xlabel('Position Level')
plt.ylabel('Salary')

# Predicting a new result with Linear Regression
lin_reg.predict(6.5)


# Predict a new result with Poly Reg
lin_reg_2.predict(poly_reg.fit_transform(6.5))
